File: octo/data-center/data-statistic/src/main/scala/com/sankuai/octo/statistic/util/CronScheduler.scala
package com.sankuai.octo.statistic.util

import org.quartz.impl.StdSchedulerFactory
import org.quartz.{JobDetail, Trigger}
import org.slf4j.LoggerFactory

/**
 * Created by wujinwu on 15/9/23.
 */
object CronScheduler {

  private val logger = LoggerFactory.getLogger(this.getClass)

  private val cronScheduler = new StdSchedulerFactory().getScheduler


  cronScheduler.start()
  // 在jvm退出时优雅关闭
  Runtime.getRuntime.addShutdownHook(new Thread() {
    override def run(): Unit = {
      cronScheduler.shutdown(true)
    }
  })

  def scheduleJob(job: JobDetail, trigger: Trigger) {
    try {
      cronScheduler.scheduleJob(job, trigger)
    } catch {
      case e: Exception => logger.error(s"scheduleJob Fail,job:$job,trigger:$trigger")
    }
  }
}

File: octo/mcc/java/config-server/src/main/java/com/sankuai/meituan/config/util/CronScheduler.java
package com.sankuai.meituan.config.util;

import org.quartz.JobDetail;
import org.quartz.Scheduler;
import org.quartz.Trigger;
import org.quartz.impl.StdSchedulerFactory;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * Created by lhmily on 07/26/2016.
 */
public class CronScheduler {
    private static Logger logger = LoggerFactory.getLogger(CronScheduler.class);

    private static Scheduler cronScheduler = null;

    static {
        try {
            cronScheduler = new StdSchedulerFactory().getScheduler();
            cronScheduler.start();
        } catch (Exception e) {
            logger.error("cannot get scheduler.", e);
        }
        // 在jvm退出时优雅关闭
        Runtime.getRuntime().addShutdownHook(new Thread() {
            @Override
            public void run() {
                try {
                    cronScheduler.shutdown(true);
                } catch (Exception e) {
                    logger.error("cronScheduler cannot exit.", e);
                }
            }
        });
    }

    public static void scheduleJob(JobDetail job, Trigger trigger) {
        try {
            cronScheduler.scheduleJob(job, trigger);
        } catch (Exception e){
            logger.error("scheduleJob Fail,job:{},trigger:{}", job, trigger);
            logger.error("scheduleJob Fail.", e);
        }
    }
}
